File: src/mindspec/docs.py
import re
import subprocess
from pathlib import Path

class DocParser:
    def __init__(self, workspace):
        self.workspace = workspace

    def parse_glossary(self):
        glossary_path = self.workspace.get_glossary_path()
        if not glossary_path.exists():
            return {}

        terms = {}
        with open(glossary_path, 'r') as f:
            content = f.read()

        # Match markdown table rows: | **Term** | [Link](Target) |
        matches = re.finditer(r'\|\s*\*\*([^*]+)\*\*\s*\|\s*\[[^\]]+\]\(([^)]+)\)\s*\|', content)
        for match in matches:
            term = match.group(1).strip()
            target = match.group(2).strip()
            terms[term] = target
        return terms

    def check_health(self, strict=False):
        docs_dir = self.workspace.get_docs_dir()
        glossary = self.parse_glossary()
        project_root = self.workspace.find_project_root()

        report = {
            "docs_dir_exists": docs_dir.exists(),
            "glossary_exists": self.workspace.get_glossary_path().exists(),
            "term_count": len(glossary),
            "broken_links": [],
            "warnings": []
        }

        for term, target in glossary.items():
            # Handle relative paths in links (strip anchors for existence check)
            path_part = target.split('#')[0]
            target_path = project_root / path_part
            if not target_path.exists():
                report["broken_links"].append(f"{term} -> {target}")

        # Domain structure checks (warn by default, error in strict mode)
        domains_dir = docs_dir / "domains"
        context_map = docs_dir / "context-map.md"
        expected_domains = ["core", "context-system", "workflow"]
        domain_files = ["overview.md", "architecture.md", "interfaces.md", "runbook.md"]

        if not domains_dir.exists():
            report["warnings"].append("docs/domains/ directory not found")
        else:
            for domain in expected_domains:
                domain_dir = domains_dir / domain
                if not domain_dir.exists():
                    report["warnings"].append(f"docs/domains/{domain}/ not found")
                else:
                    for f in domain_files:
                        if not (domain_dir / f).exists():
                            report["warnings"].append(f"docs/domains/{domain}/{f} not found")

        if not context_map.exists():
            report["warnings"].append("docs/context-map.md not found")

        # In strict mode, warnings become errors
        if strict and report["warnings"]:
            report["strict_failures"] = list(report["warnings"])

        # Beads hygiene checks
        beads_report = self.check_beads_hygiene(project_root)
        report["beads"] = beads_report

        return report

    def check_beads_hygiene(self, project_root):
        """Check Beads directory health: existence, durable state, runtime leaks."""
        beads_dir = project_root / ".beads"
        report = {
            "dir_exists": beads_dir.exists(),
            "durable_state": False,
            "tracked_runtime_artifacts": [],
        }

        if not beads_dir.exists():
            return report

        # Check for durable state files
        durable_files = ["issues.jsonl", "config.yaml", "metadata.json"]
        found_durable = [f for f in durable_files if (beads_dir / f).exists()]
        report["durable_state"] = len(found_durable) > 0
        report["durable_files_found"] = found_durable

        # Check for runtime artifacts tracked by git
        runtime_patterns = {
            "bd.sock", "daemon.lock", "daemon.log", "daemon.pid",
            "sync-state.json", "last-touched", ".local_version",
            "db.sqlite", "bd.db", "redirect", ".sync.lock",
        }
        runtime_extensions = {".db", ".db-wal", ".db-shm", ".db-journal"}

        try:
            result = subprocess.run(
                ["git", "ls-files", ".beads/"],
                capture_output=True, text=True, cwd=project_root
            )
            tracked_files = result.stdout.strip().splitlines() if result.stdout.strip() else []
            for tracked in tracked_files:
                filename = Path(tracked).name
                if filename in runtime_patterns:
                    report["tracked_runtime_artifacts"].append(tracked)
                elif any(filename.endswith(ext) for ext in runtime_extensions):
                    report["tracked_runtime_artifacts"].append(tracked)
        except FileNotFoundError:
            pass  # git not available

        return report

File: internal/doctor/docs.go
package doctor

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

// requiredDirs are directories that must exist under docs/ or project root.
var requiredDirs = []struct {
	path string // relative to project root
	name string // display name
}{
	{"docs/core", "docs/core/"},
	{"docs/domains", "docs/domains/"},
	{"docs/specs", "docs/specs/"},
	{"architecture", "architecture/"},
}

// expectedDomains and their expected files.
var expectedDomains = []string{"core", "context-system", "workflow"}
var domainFiles = []string{"overview.md", "architecture.md", "interfaces.md", "runbook.md"}

// termPattern matches glossary table rows: | **Term** | [Link](Target) |
var termPattern = regexp.MustCompile(`\|\s*\*\*([^*]+)\*\*\s*\|`)

// linkPattern extracts the target from a glossary link: [text](target)
var linkPattern = regexp.MustCompile(`\[[^\]]+\]\(([^)]+)\)`)

func checkDocs(r *Report, root string) {
	// Check required directories
	for _, d := range requiredDirs {
		p := filepath.Join(root, d.path)
		if dirExists(p) {
			r.Checks = append(r.Checks, Check{Name: d.name, Status: OK})
		} else {
			r.Checks = append(r.Checks, Check{
				Name:    d.name,
				Status:  Missing,
				Message: fmt.Sprintf("create %s directory", d.path),
			})
		}
	}

	// GLOSSARY.md
	checkGlossary(r, root)

	// context-map.md
	cmPath := filepath.Join(root, "docs", "context-map.md")
	if fileExists(cmPath) {
		r.Checks = append(r.Checks, Check{Name: "docs/context-map.md", Status: OK})
	} else {
		r.Checks = append(r.Checks, Check{
			Name:    "docs/context-map.md",
			Status:  Missing,
			Message: "create docs/context-map.md",
		})
	}

	// Domain subdirectory checks
	checkDomains(r, root)
}

func checkGlossary(r *Report, root string) {
	glossaryPath := filepath.Join(root, "GLOSSARY.md")
	data, err := os.ReadFile(glossaryPath)
	if err != nil {
		r.Checks = append(r.Checks, Check{
			Name:    "GLOSSARY.md",
			Status:  Missing,
			Message: "create GLOSSARY.md in project root",
		})
		return
	}

	content := string(data)

	// Count terms (lines matching | **Term** |)
	terms := termPattern.FindAllString(content, -1)
	r.Checks = append(r.Checks, Check{
		Name:    "GLOSSARY.md",
		Status:  OK,
		Message: fmt.Sprintf("(%d terms)", len(terms)),
	})

	// Check for broken links
	lines := strings.Split(content, "\n")
	var broken []string
	for _, line := range lines {
		if !termPattern.MatchString(line) {
			continue
		}
		linkMatch := linkPattern.FindStringSubmatch(line)
		if linkMatch == nil {
			continue
		}
		target := linkMatch[1]
		// Strip anchor for file existence check
		pathPart := strings.SplitN(target, "#", 2)[0]
		if pathPart == "" {
			continue
		}
		fullPath := filepath.Join(root, pathPart)
		if !fileExists(fullPath) {
			// Extract term name
			termMatch := termPattern.FindStringSubmatch(line)
			term := ""
			if termMatch != nil {
				term = strings.TrimSpace(termMatch[1])
			}
			broken = append(broken, fmt.Sprintf("%s -> %s", term, target))
		}
	}

	if len(broken) > 0 {
		r.Checks = append(r.Checks, Check{
			Name:    "Glossary links",
			Status:  Error,
			Message: fmt.Sprintf("%d broken: %s", len(broken), strings.Join(broken, "; ")),
		})
	} else {
		r.Checks = append(r.Checks, Check{
			Name:    "Glossary links",
			Status:  OK,
			Message: "all links verified",
		})
	}
}

func checkDomains(r *Report, root string) {
	domainsDir := filepath.Join(root, "docs", "domains")
	if !dirExists(domainsDir) {
		return // already reported as missing in requiredDirs
	}

	for _, domain := range expectedDomains {
		domainDir := filepath.Join(domainsDir, domain)
		if !dirExists(domainDir) {
			r.Checks = append(r.Checks, Check{
				Name:    fmt.Sprintf("docs/domains/%s/", domain),
				Status:  Warn,
				Message: "domain directory not found",
			})
			continue
		}

		for _, f := range domainFiles {
			fp := filepath.Join(domainDir, f)
			name := fmt.Sprintf("docs/domains/%s/%s", domain, f)
			if fileExists(fp) {
				r.Checks = append(r.Checks, Check{Name: name, Status: OK})
			} else {
				r.Checks = append(r.Checks, Check{
					Name:    name,
					Status:  Warn,
					Message: "file not found",
				})
			}
		}
	}
}

func dirExists(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.IsDir()
}

func fileExists(path string) bool {
	info, err := os.Stat(path)
	return err == nil && !info.IsDir()
}
